src/v/utils/token_bucket.h

```diff
@@ -117,7 +117,7 @@ class token_bucket {
 
         if (_rate > new_rate) {
             if (_sem.current() <= _rate) {
-                _sem.consume(_rate - new_rate);
+                _sem.consume(std::min(_rate - new_rate, _sem.current()));
```

Comment:
I think that old behavior is correct. If the TB got throttled we need to wait for it to be refilled and we should also take into account the previous state. 

Example. The TB was set to 100MiB/s and the clients consumed around 10MiB/s. There was no throttling. The `_sem.current()` has around 90M units. Now we're changing the TB rate limit to 1MiB/s. In this case the waiter should be throttled for around 9s to proceed. The old code will unconditionally consume 99M units from the TB. The waiters will be throttled for 9s. The new code will consume from the `_sem` 90M units unconditionally.  So the waiters will consume from the semaphore after shorter period. But this doesn't look correct to me because they should be throttled because the new limit is 1MiB/s and they are trying to consume 10MiB/s. With new logic there will be a spike in consumption. The waiters will be able to use more than 1MiB/s for a short brief and then the consumption will be pinned to 1MiB/s. The old logic will pin consumption to 1MiB/s right away.

It all depends on what you want. But the current implementation has the most vanilla TB behavior you can think about.

Reply:
My point is that the 10M tokens that were consumed shouldn't penalize future users. The caller _had the right to them_ because they were consumed under a policy was 100M/s.

In the next second, the policy is 1M/s. Why it does make sense to penalize the callers for 9 seconds and don't give them any tokens? It does not make sense to me.

_The requested policy is 1M/s but the effective policy for the next 10 seconds will be actually 0.1M/s._

Reply:
How about the following example.

Imagine a real bucket. You have a water refill rate and water consume rate.

The bucket starts full.

You consume from it.

If the bucket becomes empty at any point, you can't consuming any more from it. Instead you waiting for it to refill.

Our token bucket works exactly like this. At no point you are allowed to consume non-existing tokens.

Now, imagine that the water refill rate decreased. Whatever was in the bucket stays in the bucket. You are not saying to the consumer "oh hey wait you did consume too fast recently so no water for you". The bucket was being consumed at a time where refill rate was greater.

At most, it makes sense to empty the bucket as otherwise it would allow out-of-current-policy bursting.

Reply:
The reason why I made this PR is that in integrations tests (and real world too) if I have a limit of 1 GiB/s and I decrease it to 50 MiB/s for tiered storage downloads then the system halts (!) for good 20 seconds. The behavior in this practical example does not make sense to me. The system should never halt for that long.

Reply:
> Now, imagine that the water refill rate decreased. Whatever was in the bucket stays in the bucket.

This is not what the change in the code does. It empties the bucket when the rate decreases.

> The behavior in this practical example does not make sense to me. The system should never halt for that long.

This is why I wrote `It all depends on what you want`. Because when system is running with 1GiB/s workload and you're resetting the rate to 1MiB/s some throttling has to be applied in both cases. It only looks unnatural if your workload also changes with the rate (I guess this is what happens in the integration test).  I guess that I don't have any practical concern and both options will work. Maybe this could be controlled by the additional parameter for the TB reset method.